utils/resb/src/text/reader/parse_state.rs

```diff
@@ -0,0 +1,194 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::{
+    cell::RefCell,
+    ops::{Deref, Range, RangeFrom, RangeTo},
+    rc::Rc,
+    str::{CharIndices, Chars},
+};
+
+use indexmap::IndexSet;
+use nom::{Compare, FindSubstring, InputIter, InputLength, InputTake, Slice, UnspecializedInput};
+
+use crate::bundle::Key;
+
+/// The `ParseState` struct contains the information required to manage the
+/// parse at any given position in the text.
+///
+/// # Caveats
+///
+/// This struct will be cloned or split repeatedly throughout the parse and so
+/// should not contain excess state. Code making use of the parse state should
+/// ensure that only a few exist at a time.
+#[derive(Clone, Debug)]
+pub(in crate::text::reader) struct ParseState<'a> {
+    /// The current position in the input, represented as a slice beginning at
+    /// the next character to be read and ending at the end of input.
+    input: &'a str,
+
+    /// The keys present in the resource bundle in the order we encounter them
+    /// during the parse.
+    ///
+    /// This isn't important to the in-memory representation of resource bundles
+    /// but is a part of the binary writing algorithm in the C++ parser and so
+    /// we keep track of it in order to keep strict compatibility in output
+    /// files.
+    ///
+    /// This cell should only be mutated internally when:
+    /// - Adding a key while parsing a table resource.
+    /// - Taking the list of keys when the parse is completed.
+    keys_in_discovery_order: Rc<RefCell<IndexSet<Key<'a>>>>,
```

Comment:
why do you use `Rc<RefCell>>` here? Why not just expose the `IndexSet`?

Reply:
This is probably a relic of an earlier stage of development. It no longer seems to be necessary.

Reply:
I've gone back to take another look at this and the reason I'm using this is that `nom` inputs are created and cloned frequently during the parsing process. While it seems to be possible to get away with keeping the `IndexSet` bare, we'd incur a cost in cloning.